### ISSUE TITLE
Documentation: How to use a promise-based Environment?  (e.g. replacing react-relay-network-layer authentication)

### ISSUE THREAD

**Issue by lukecwilliams** [facebook/relay#1781]
In Relay Classic, we just pass `authMiddleware` to `react-relay-network-layer` and the token is properly refreshed.

```
authMiddleware({
    token: () => accessToken,
    tokenRefreshPromise: Lock,
}),
```

What's the equivalent of this in Relay Modern?  We need a solution that allows deferred creation of Environment, or rendering with a new Environment in the callback on the authentication call.  The documentation references using multiple Environments, but I'm not aware of any examples.

Thank you!

**Comment by sibelius** [comment]
it is much easier to do this on Relay modern

just handle this in your `fetchQuery` function

https://github.com/sibelius/ReactNavigationRelayModern/blob/master/src/createRelayEnvironment.js#L13

**Comment by lukecwilliams** [comment]
@sibelius I think my problem is that another fetch is firing before the first fetch completes, which gets us into a mixed-up state.

I have this call in `async function fetchQuery` which either fetches the existing authToken or goes to the auth server for one:

    const authToken = await AuthToken.get();

But if another fetch is triggered before auth function returns, the authentication happens all over again which breaks our app.

The answer should be something like, "the second call to fetchQuery also waits on the same auth call already in progress," but I haven't been able to figure out how to do that with promises.  May be a JS/async programming pattern I don't know about yet - do you have any guidance?

We could also fix this if Relay was not firing two calls in quick succession at app start, a separate issue, however I'd like to implement a more robust solution.

**Comment by lukecwilliams** [comment]
I solved this with Promise.resolve(), which lets multiple places in the code wait on the result of a single promise.

Detail: 
- fetchQuery gets an auth token (JWT) with:

      const authToken = await AuthToken.get();

AuthToken looks like:

    class AuthToken {
      private static _accessTokenPromise: Promise<string>;
      public static async get() {
        if (!this._accessTokenPromise)
          this._accessTokenPromise = this.AuthFunction(); // AuthFunction returns a promise

        return await Promise.resolve(this._accessTokenPromise);
      }
    }

I'm not sure this is the best solution (def open to comments), nor how well it will hold up when we're testing and doing dependency injection, but for right now the static Promise.resolve() method solved the issue.
